src/magic_folder/tahoe_client.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+import json
+
+from twisted.internet.defer import (
+    inlineCallbacks,
+    returnValue,
+)
+from hyperlink import (
+    DecodedURL,
+)
+
+import attr
+
+
+@attr.s
+class TahoeClient(object):
+    """
+    An object that knows how to call a particular tahoe client's
+    WebAPI. Usually this means a node-directory (to get the base URL)
+    and a treq client (to make HTTP requests).
+    """
+
+    # node_directory = attr.ib()
+    url = attr.ib()
+    http_client = attr.ib()
+
+    @inlineCallbacks
+    def create_immutable_directory(self, directory_data):
+        post_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"t", u"mkdir-immutable")],
+        )
+        res = yield self.http_client.post(
+            post_uri.to_text(),
+            json.dumps(directory_data),
+        )
+        capability_string = yield res.content()
+        returnValue(
+            capability_string.strip()
+        )
+
+    @inlineCallbacks
+    def create_immutable(self, producer):
+        """
+        :param producer: can take anything that treq's data= method to
+            treq.request allows which is currently: str, file-like or
+            IBodyProducer. See
+            https://treq.readthedocs.io/en/release-20.3.0/api.html#treq.request
+        """
+
+        put_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"mutable", u"false")],
+        )
+        res = yield self.http_client.put(
+            put_uri.to_text(),
+            data=producer,
+        )
+        capability_string = yield res.content()
+        returnValue(
+            capability_string.strip()
+        )
+
+    @inlineCallbacks
+    def download_capability(self, cap):
+        get_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"uri", cap.decode("ascii"))],
+        )
+        res = yield self.http_client.get(get_uri.to_text())
+        data = yield res.content()
+        returnValue(data)
+
+    @inlineCallbacks
+    def stream_capability(self, cap, filelike):
+        get_uri = self.url.replace(
+            path=(u"uri",),
+            query=[(u"uri", cap.decode("ascii"))],
+        )
+        res = yield self.http_client.get(get_uri.to_text())
+        yield res.collect(filelike.write)
+
+
+@inlineCallbacks
+def create_tahoe_client(url, http_client):
+    """
+    Create a new TahoeClient instance that is speaking to a particular
+    Tahoe node.
+
+    :param url: the base URL of the Tahoe instance
+
+    :param http_client: a Treq HTTP client
+
+    :returns: a TahoeClient instance
+    """
+
+    client = TahoeClient(
+        url=DecodedURL.from_text(url),
+        http_client=http_client,
+    )
+    yield  # maybe we want to at least try getting / to see if it's alive?
```

Comment:
I think this function should be completely synchronous.  A Tahoe client is completely described by the API root to use.  Whether there is a functioning node serving an API at that root is a separate question.  The client might have some methods for interrogating the health of the node or something but it doesn't seem like you should have to asynchronously loop around `create_tahoe_client` waiting for the node to behave nicely - when the node could fall over a split second after `create_tahoe_client` finally decides it's okay.  Instead, you have to be prepared for failure at every point of interaction and add your error handling logic there.

Given this, and similar to comments left elsewhere about `LocalAuthor`, I wonder what the benefit of having both `TahoeClient` and `create_tahoe_client` is.  If `TahoeClient` were private then `create_tahoe_client` would make sense to me (hide the type) but since it is public `create_tahoe_client` seems redundant.

Shall we try to say something general - like "types should be hidden from the public interface"?  This means we need explicit interfaces for a bunch more things since saying things like `":return: A `_TahoeClient` instance" is a bit less cool.

Reply:
"In general" I use `create_*()` style functions by default now because there's "often" stuff to do to set up an object and it seems way more consistent to just always have "create" functions. Probably I guessed we'd add more to this at some point? (Like load information from Tahoe's node_directory?)

Yes, "check if it's alive" might be a bad idea but also maybe it could be used for better user-feedback (like "I can't reach `localhost:1234` at all" versus N retries and then something less-scrutable like "I couldn't download capability X because of reasons"). Maybe the caller can do that .. but so, hedging the bets here on sync vs async. And for context, I'd really rather use "async def" for things that "might be allowed to do async things" but we can't have that now so I'm tending to use `@inlineCallbacks` for that kind of "marking" instead.

Given that going from sync->async is the "hard way", are we 100% sure this will never ever want to do any async work at all? Are we 100% sure that we will never want to do any "setup work" (of the kind that probably shouldn't be in `__init__`, like possibly loading tahoe-config stuff)? Perhaps there's a risk of "over-future-proofing" here, but I think choosing "async" is good  (and it's way easier to go back to sync later if it proves that we *don't* need anything fancier than this in the near future).